[Quant][fx] Decouple prepare_*fx from training/eval modes

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75401

Summary: This commit removes asserts that require prepare_fx to
be run in eval mode and prepare_qat_fx to be run in training mode.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_prepare_mode

Reviewers: jerryzh168

Subscribers: jerryh168

Differential Revision: [D35457100](https://our.internmc.facebook.com/intern/diff/D35457100)

andrewor14 committed:
[Quant][fx] Decouple prepare_*fx from training/eval modes

Summary: This commit removes asserts that require prepare_fx to
be run in eval mode and prepare_qat_fx to be run in training mode.

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers: jerryzh168

Subscribers: jerryh168

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35457100).

jerryzh168 commented on torch/ao/quantization/fuser_method_mappings.py:
```diff
@@ -142,7 +136,6 @@ def fuse_convtranspose_bn(is_qat, convt, bn):
 
     if is_qat:
-        assert convt.training, "qat is only supported when convt.training is True currently"
         raise Exception("Fusing ConvTranspose+BatchNorm not yet supported in training.")
```
we can update the exception message as well, we can say it does not support fusion for quantization aware training

jerryzh168 reviewed (approved):
thanks! please add a small test to check this works in TestQuantizeFx as well

andrewor14 committed:
Update on "[Quant][fx] Decouple prepare_*fx from training/eval modes"


Summary: This commit removes asserts that require prepare_fx to
be run in eval mode and prepare_qat_fx to be run in training mode.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_prepare_mode

Reviewers: jerryzh168

Subscribers: jerryh168

Differential Revision: [D35457100](https://our.internmc.facebook.com/intern/diff/D35457100)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35457100).
